Update run_xnli.py to use Datasets library

Qbiwan opened the pull request:
# What does this PR do?


<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #9754


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@sgugger 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @jplu

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- nlp datasets: [different repo](https://github.com/huggingface/nlp)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

Qbiwan committed:
Merge pull request #1 from huggingface/master

update my fork with lastest

Qbiwan committed:
Merge remote-tracking branch 'upstream/master'

Qbiwan committed:
Merge remote-tracking branch 'upstream/master'

Qbiwan committed:
remove xnli_compute_metrics, add load_dataset, load_metric, set_seed,metric.compute,load_metric

Qbiwan committed:
remove xnli_compute_metrics, add load_dataset, load_metric, set_seed,metric.compute,load_metric

Qbiwan committed:
Merge remote-tracking branch 'upstream/master'

patrickvonplaten committed:
fix led not defined (#9828)

patil-suraj commented on examples/text-classification/run_xnli.py:
```diff
@@ -303,44 +388,50 @@ def evaluate(args, model, tokenizer, prefix=""):
 
-def load_and_cache_examples(args, task, tokenizer, evaluate=False):
-    if args.local_rank not in [-1, 0] and not evaluate:
+def load_and_cache_examples(model_args, data_args, training_args, task, tokenizer, evaluate=False):
```
Thank you for working on this!

But it seems that the dataset is still loaded using the old way without using the `datasets` library.

Also IMO it would be better, if we update the script to use `Trainer`, like `run_glue.py`, let us know if you want to work on this, happy to help here :)

Qbiwan replied:
I see, use the Trainer() let me see if I could figure it out. thanks for the tip @patil-suraj

Qbiwan replied:
@patil-suraj I loaded dataset at line 501 in `main()` using `load_dataset(data_args.task_name, model_args.language, data_dir=data_args.data_dir)` but only used `data_dir` here to strictly to cache `cached_feature_file`. Am I still downloading the dataset `xnli` incorrectly?

Qbiwan replied:
Hi @patil-suraj  I took another crack at the PR. I tried to adapt `run_glue.py` as closely to the task of `fine-tuning mBERT (multi-lingual BERT) on the XNLI dataset` as best as I could. As these options below weren't in the original run_xnli.py, I did not include them:
- option to use csv or json train/test files from user
- do_predict   

I also assumed that this script is solely to run xnli task. Hopefully it's closer to what you required :)
Let me know if there is anything you wish to change.

abhishekkrthakur committed:
ConvBERT Model (#9717)

* finalize convbert

* finalize convbert

* fix

* fix

* fix

* push

* fix

* tf image patches

* fix torch model

* tf tests

* conversion

* everything aligned

* remove print

* tf tests

* fix tf

* make tf tests pass

* everything works

* fix init

* fix

* special treatment for sepconv1d

* style

* 🙏🏽

* add doc and cleanup

* add electra test again

* fix doc

* fix doc again

* fix doc again

* Update src/transformers/modeling_tf_pytorch_utils.py

Co-authored-by: Lysandre Debut <lysandre@huggingface.co>

* Update src/transformers/models/conv_bert/configuration_conv_bert.py

Co-authored-by: Lysandre Debut <lysandre@huggingface.co>

* Update docs/source/model_doc/conv_bert.rst

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Update src/transformers/models/auto/configuration_auto.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Update src/transformers/models/conv_bert/configuration_conv_bert.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* conv_bert -> convbert

* more fixes from review

* add conversion script

* dont use pretrained embed

* unused config

* suggestions from julien

* some more fixes

* p -> param

* fix copyright

* fix doc

* Update src/transformers/models/convbert/configuration_convbert.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

* comments from reviews

* fix-copies

* fix style

* revert shape_list

Co-authored-by: Lysandre Debut <lysandre@huggingface.co>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

patrickvonplaten committed:
[Setup.py] update jaxlib (#9831)

* update jaxlib

* Update setup.py

* update table

jplu committed:
Add a test for mixed precision (#9806)

sgugger committed:
Setup logging with a stdout handler (#9816)

jncasey committed:
Fix auto-resume training from checkpoint (#9822)

* Fix auto-resume training from checkpoint

* style fixes

patrickvonplaten committed:
fix typo with mt5 init (#9830)

Narsil committed:
Adding a test to prevent late failure in the Table question answering (#9808)

pipeline.

- If table is empty then the line that contain `answer[0]` will fail.
- This PR add a check to prevent `answer[0]`.
- Also adds an early check for presence of `table` and `query` to
prevent late failure and give better error message.
- Adds a few tests to make sure these errors are correctly raised.

jplu committed:
Remove a TF usage warning and rework the documentation (#9756)

* Rework documentation

* Update the template

* Trigger CI

* Restore the warning but with the TF logger

* Update convbert doc

tomohideshibata committed:
Delete a needless duplicate condition (#9826)

Co-authored-by: Tomohide Shibata <tomshiba@yahoo-corp.jp>

jplu committed:
Clean TF Bert (#9788)

* Start cleaning BERT

* Clean BERT and all those depends of it

* Fix attribute name

* Apply style

* Apply Sylvain's comments

* Apply Lysandre's comments

* remove unused import

sgugger committed:
Add a flag for find_unused_parameters (#9820)

* Add a flag for find_unused_parameters

* Apply suggestions from code review

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

* Remove negation

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

sgugger commented on examples/text-classification/run_xnli.py:
```diff
@@ -36,10 +37,12 @@
 from transformers import glue_convert_examples_to_features as convert_examples_to_features
-from transformers import xnli_compute_metrics as compute_metrics
+from transformers import set_seed
 from transformers import xnli_output_modes as output_modes
 from transformers import xnli_processors as processors
```
Those shouldn't be imported and all preprocessing should be done in the example script, like for the new `run_glue`.

Qbiwan replied:
ah i see; let me take another crack at it. If it proves to be too advanced for me, I'll just close the PR.

sgugger replied:
Sure! And happy to help if you have some issues.

sgugger reviewed (changes requested):
Thanks for your PR! As @patil-suraj mentioned it, this doesn't fulfill the goal set by the issue: the new script should not rely on the Transformers library for the data preprocessing (those functions will be removed from the library in a future version) but expose all preprocessing in the script, like it's done for `run_glue`.

jplu committed:
Fix template (#9840)

LysandreJik committed:
Fix model templates (#9842)

kiyoungkim1 committed:
Add tpu_zone and gcp_project in training_args_tf.py (#9825)

* add tpu_zone and gcp_project in training_args_tf.py

* make style

Co-authored-by: kykim <kykim>

LysandreJik committed:
Labeled pull requests (#9849)

LysandreJik committed:
Test (#9851)

sgugger committed:
When resuming training from checkpoint, Trainer loads model (#9818)

* Whenresuming training from checkpoint, Trainer loads model

* Finish cleaning tests

* Address review comment

* Use global_step from state

sgugger committed:
Allow --arg Value for booleans in HfArgumentParser (#9823)

* Allow --arg Value for booleans in HfArgumentParser

* Update last test

* Better error message

stas00 committed:
[traner] fix --lr_scheduler_type choices (#9800)

* fix --lr_scheduler_type choices

* rewrite to fix for all enum-based cl args

* cleanup

* adjust test

* style

* Proposal that should work

* Remove needless code

* Fix test

Co-authored-by: Sylvain Gugger <sylvain.gugger@gmail.com>

stefan-it committed:
ADD BORT (#9813)

* tests: add integration tests for new Bort model

* bort: add conversion script from Gluonnlp to Transformers 🚀

* bort: minor cleanup (BORT -> Bort)

* add docs

* make fix-copies

* clean doc a bit

* correct docs

* Update docs/source/model_doc/bort.rst

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Update docs/source/model_doc/bort.rst

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* correct dialogpt doc

* correct link

* Update docs/source/model_doc/bort.rst

* Update docs/source/model_doc/dialogpt.rst

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

* make style

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

abhishekkrthakur committed:
Pin memory in Trainer by default (#9857)

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

LysandreJik committed:
Partial local tokenizer load (#9807)

* Allow partial loading of a cached tokenizer

* Warning > Info

* Update src/transformers/tokenization_utils_base.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Raise error if not local_files_only

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

LysandreJik committed:
Remove submodule (#9868)

Narsil committed:
Fixing flaky conversational test + flag it as a pipeline test. (#9837)

mfuntowicz committed:
Fix computation of attention_probs when head_mask is provided. (#9853)

* Fix computation of attention_probs when head_mask is provided.

Signed-off-by: Morgan Funtowicz <funtowiczmo@gmail.com>

* Apply changes to the template

Co-authored-by: Lysandre <lysandre.debut@reseau.eseo.fr>

sgugger committed:
Deprecate model_path in Trainer.train (#9854)

joeddav committed:
tutorial typo

stancld committed:
Remove redundant `test_head_masking  = True` flags in test files (#9858)

* Remove redundant test_head_masking = True flags

* Remove all redundant test_head_masking flags in PyTorch test_modeling_* files

* Make test_head_masking = True as a default choice in test_modeling_tf_commong.py

* Remove all redundant test_head_masking flags in TensorFlow
test_modeling_tf_* files

* Put back test_head_masking=False fot TFT5 models

stas00 committed:
[docs] expand install instructions (#9817)

* expand install instructions

* fix

* white space

* rewrite as discussed in the PR

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* change the wording to encourage issue report

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

abhishekkrthakur committed:
on_log event should occur *after* the current log is written (#9872)

abhishekkrthakur committed:
pin_memory -> dataloader_pin_memory (#9874)

Narsil committed:
Adding a new `return_full_text` parameter to TextGenerationPipeline. (#9852)

* Adding a new `return_full_text` parameter to TextGenerationPipeline.

For text-generation, it's sometimes used as prompting text.
In that context, prefixing `generated_text` with the actual input
forces the caller to take an extra step to remove it.

The proposed change adds a new parameter (for backward compatibility).
`return_full_text` that enables the caller to prevent adding the prefix.

* Doc quality.

ethch18 committed:
Clarify use of unk_token in tokenizer docstrings (#9875)

jplu committed:
Add XLA test (#9848)

sgugger committed:
When on sagemaker use their env variables for saves (#9876)

* When on sagemaker use their env variables for saves

* Address review comments

* Quality

stas00 committed:
correctly handle mt5 (#9879)

stas00 committed:
refactor deepspeed setup devices (#9880)

stas00 committed:
[doc] nested markup is invalid in rst (#9898)

Apparently nested markup in RST is invalid: https://docutils.sourceforge.io/FAQ.html#is-nested-inline-markup-possible

So currently this line doesn't get rendered properly, leaving inner markdown unrendered, resulting in:
```
https://docutils.sourceforge.io/FAQ.html#is-nested-inline-markup-possible
```

This PR removes the bold which fixes the link.

Qbiwan committed:
Merge remote-tracking branch 'upstream/master'

Qbiwan committed:
Merge branch 'master' of https://github.com/Qbiwan/transformers into dataclass_xnli_2nd_pass

Qbiwan committed:
merge conflict run_xnli

Qbiwan committed:
amend README

patil-suraj commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,325 @@
-    optimizer = AdamW(optimizer_grouped_parameters, lr=args.learning_rate, eps=args.adam_epsilon)
-    scheduler = get_linear_schedule_with_warmup(
-        optimizer, num_warmup_steps=args.warmup_steps, num_training_steps=t_total
+    task_name: Optional[str] = field(
```
Since the script is written specifically for `xnli` I would remove the task_name argument, as we already know the task

Qbiwan replied:
I removed `task_name` from `DataTrainingArguments` 
https://github.com/huggingface/transformers/blob/c2b78bc803e6468771cdca0c47cd03040f7751dc/examples/text-classification/run_xnli.py#L52-L79

patil-suraj commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,325 @@
 
-    dataset = TensorDataset(all_input_ids, all_attention_mask, all_token_type_ids, all_labels)
-    return dataset
+    def __post_init__(self):
+        if self.task_name is not None:
+            self.task_name = self.task_name.lower()
+            if self.task_name not in task_to_keys.keys():
+                raise ValueError("Unknown task, you should pick one in " + ",".join(task_to_keys.keys()))
```
We could remove this method, as said above we already know the task

Qbiwan replied:
I removed __post_init__

patil-suraj commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,325 @@
-        required=True,
-        help="The output directory where the model predictions and checkpoints will be written.",
+    train_language: Optional[str] = field(
+        default=None, metadata={"help": "Train language if is different of the evaluation language."}
```
```suggestion
        default=None, metadata={"help": "Train language if it is different from the evaluation language."}
```

Qbiwan replied:
I amended the sentence
https://github.com/huggingface/transformers/blob/c2b78bc803e6468771cdca0c47cd03040f7751dc/examples/text-classification/run_xnli.py#L94-L96

patil-suraj commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,325 @@
+    # download the dataset.
+    if data_args.task_name is not None:
+        # Downloading and loading xnli dataset from the hub.
+        datasets = load_dataset(data_args.task_name, model_args.language)
```
Here the language for train and validation dataset can be different, so we should load the train and validation dataset separately,  to load a specific split you pass the `split` argument to `load_dataset`

```python
train_dataset = datasets.load_dataset("xnli", train_language, split="train")
eval_dataset = datasets.load_dataset("xnli", language, split="validation")
```

Qbiwan replied:
changed `load_dataset` to load separately for `train_dataset` and `eval_dataset`
https://github.com/huggingface/transformers/blob/c2b78bc803e6468771cdca0c47cd03040f7751dc/examples/text-classification/run_xnli.py#L189-L193

patil-suraj commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,325 @@
+        result = tokenizer(*args, padding=padding, max_length=data_args.max_seq_length, truncation=True)
+        return result
+
+    datasets = datasets.map(preprocess_function, batched=True, load_from_cache_file=not data_args.overwrite_cache)
+
+    train_dataset = datasets["train"]
+    eval_dataset = datasets["validation"]
```
Same here, we should process the train and eval dataset separately

Qbiwan replied:
I've applied `preprocess_function` on train and eval separately

https://github.com/huggingface/transformers/blob/c2b78bc803e6468771cdca0c47cd03040f7751dc/examples/text-classification/run_xnli.py#L244-L250

I've also removed the conditional `if data_args.task_name is not None` in `compute_metrics()`, 

https://github.com/huggingface/transformers/blob/19233fed01e66dc50df88c8b5e9c7adeef25f064/examples/text-classification/run_xnli.py#L268-L275

and changed it to the following:
https://github.com/huggingface/transformers/blob/c2b78bc803e6468771cdca0c47cd03040f7751dc/examples/text-classification/run_xnli.py#L260-L266

patil-suraj commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,325 @@
         num_labels=num_labels,
-        finetuning_task=args.task_name,
-        cache_dir=args.cache_dir,
+        finetuning_task=data_args.task_name,
```
```suggestion
        finetuning_task="xnli",
```

Qbiwan replied:
I've changed it to reference the variable `task_name`
https://github.com/huggingface/transformers/blob/c2b78bc803e6468771cdca0c47cd03040f7751dc/examples/text-classification/run_xnli.py#L201-L205

patil-suraj reviewed (changes requested):
Great Job! I left a few comments

Qbiwan committed:
removed data_args.task_name and replaced with task_name = "xnli"; use split function to load train and validation dataset separately; remove __post_init__; remove flag --task_name from README.

sgugger commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,315 @@
-except ImportError:
-    from tensorboardX import SummaryWriter
 
+task_to_keys = {"xnli": ("premise", "hypothesis")}
```
I think we can just use `"premise"` and `"hypothesis"` later on, since there is only one task we don't need this variable anymore.

Qbiwan replied:
I've removed dict `task_to_keys` as we can later refer directly to "xnli", "premise", and "hypothesis" 
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L237-L238
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L204

https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L314
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L318
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L256
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L186-L188
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L190

sgugger commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,315 @@
+    # or by passing the --help flag to this script.
+    # We now keep distinct sets of args, for a cleaner separation of concerns.
+
+    task_name = "xnli"
```
Same for this one, we can just use "xnli" later on.

Qbiwan replied:
I've removed `task_name` and used "xnli" driectly 
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L204

https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L314
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L318
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L256
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L186-L188
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L190

sgugger commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,315 @@
-    if args.local_rank not in [-1, 0]:
-        torch.distributed.barrier()  # Make sure only the first process in distributed training will download model & vocab
+    if training_args.local_rank not in [-1, 0]:
+        torch.distributed.barrier()  # Make sure only the first process in distributed training will download model & vocabr
```
```suggestion
        torch.distributed.barrier()  # Make sure only the first process in distributed training will download model & vocab
```

Qbiwan replied:
I've amended the typo

https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L197

sgugger commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,315 @@
         num_labels=num_labels,
-        finetuning_task=args.task_name,
-        cache_dir=args.cache_dir,
+        finetuning_task=task_name,
```
```suggestion
        finetuning_task="xnli",
```

Qbiwan replied:
I've replaced task_name with str "xnli"
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L201-L205

sgugger commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,315 @@
-    model.to(args.device)
+    def preprocess_function(examples):
+        # Tokenize the texts
+        args = (
+            (examples[sentence1_key],) if sentence2_key is None else (examples[sentence1_key], examples[sentence2_key])
+        )
+        result = tokenizer(*args, padding=padding, max_length=data_args.max_seq_length, truncation=True)
```
```suggestion
        result = tokenizer(examples["premise"], examples["hypothesis"], padding=padding, max_length=data_args.max_seq_length, truncation=True)
```
There is always a second sentence here, so need for the args line.

Qbiwan replied:
I've added the second sentence and remove the unnecessary `sentence1_key` and `sentence2_key`

https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L234-L242

sgugger commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,315 @@
-    if args.local_rank == 0:
-        torch.distributed.barrier()  # Make sure only the first process in distributed training will download model & vocab
+    # Preprocessing the datasets
+    sentence1_key, sentence2_key = task_to_keys[task_name]
```
This line can be removed.

Qbiwan replied:
I've remove the unnecessary `sentence1_key` and `sentence2_key`

sgugger commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,315 @@
+        preds = p.predictions[0] if isinstance(p.predictions, tuple) else p.predictions
+        preds = np.argmax(preds, axis=1)
+        result = metric.compute(predictions=preds, references=p.label_ids)
+        if len(result) > 1:
+            result["combined_score"] = np.mean(list(result.values())).item()
```
What is the metric for xnli? Depending on it we can either remove the two lines (if there is just one) or remove the test.

Qbiwan replied:
The metric for xnli is just the simple accuracy,
```
>>> from datasets import load_metric
>>> metric = load_metric("xnli")
>>> metric.compute
<bound method compute of Metric(name: "xnli", features: {'predictions': Value(dtype='int64', id=None), 'references': Value(dtype='int64', id=None)}, usage: """
Computes XNLI score which is just simple accuracy.
Args:
    predictions: Predicted labels.
    references: Ground truth labels.
Returns:
    'accuracy': accuracy
```
hence I've removed the test for `len(result)>1` and amended `compute_metrics()` to

https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L260-L263

sgugger reviewed (approved):
This is in good shape to be merged, thanks a lot for your work! I just have a few comments on how to simplify things here and there since there is only one task to deal with in the new script.

One question I have is, is the tokenizer the same for the training and evaluation datasets, even if the languages, can be different?

patil-suraj commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,315 @@
+    # download the dataset.
+
+    # Downloading and loading xnli dataset from the hub.
+    train_dataset = load_dataset(task_name, model_args.train_language, split="train")
```
`train_language` can be `None`, so we need a check here. If it's `None`,  set it to `model_args.language`.

Qbiwan replied:
I've set train_dataset to take in `model_args.language` when `train_language` is None, else `model_args.train_language` 
https://github.com/huggingface/transformers/blob/0f6858f34ff5b56a451f893548269904f3efb751/examples/text-classification/run_xnli.py#L184-L190

Qbiwan committed:
removed dict task_to_keys, use str "xnli" instead of variable task_name, change preprocess_function to use examples["premise"], examples["hypothesis"] directly, remove sentence1_key and sentence2_key, change compute_metrics function to cater only to accuracy metric, add condition for train_langauge is None when using dataset.load_dataset()

sgugger commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,312 @@
 
     # Load pretrained model and tokenizer
-    if args.local_rank not in [-1, 0]:
+    if training_args.local_rank not in [-1, 0]:
         torch.distributed.barrier()  # Make sure only the first process in distributed training will download model & vocab
```
We don't need those anymore actually (sorry to have missed it before). The `from_pretrained` method does all the work for us.

Qbiwan replied:
I've removed from the script
`import torch` 
and 
`
    if training_args.local_rank not in [-1, 0]:
        torch.distributed.barrier()  # Make sure only the first process in distributed training will download model & vocab
`

sgugger commented:
Just tested the script locally and it seems to work great, congrats! We are almost done! The last part would be top adapt the end of the README in the text-classification folder to reflect how to use the new script (since the arguments are a bit different).

Qbiwan committed:
removed `torch.distributed.barrier()` and `import torch` as `from_pretrained` is able to do the work; amend README

Qbiwan commented:
> Just tested the script locally and it seems to work great, congrats! We are almost done! The last part would be top adapt the end of the README in the text-classification folder to reflect how to use the new script (since the arguments are a bit different).

I've changed the script in README from
```
export XNLI_DIR=/path/to/XNLI

python run_xnli.py \
  --model_name_or_path bert-base-multilingual-cased \
  --language de \
  --train_language en \
  --do_train \
  --do_eval \
  --data_dir $XNLI_DIR \
  --per_device_train_batch_size 32 \
  --learning_rate 5e-5 \
  --num_train_epochs 2.0 \
  --max_seq_length 128 \
  --output_dir /tmp/debug_xnli/ \
  --save_steps -1
```
to 
```
python run_xnli.py \
  --model_name_or_path bert-base-multilingual-cased \
  --language de \
  --train_language en \
  --do_train \
  --do_eval \
  --per_device_train_batch_size 32 \
  --learning_rate 5e-5 \
  --num_train_epochs 2.0 \
  --max_seq_length 128 \
  --output_dir /tmp/debug_xnli/ \
  --save_steps -1
```

I've also removed these sentences below from [Fine-tuning on XNLI](https://github.com/huggingface/transformers/blob/master/examples/text-classification/README.md#fine-tuning-on-xnli)
```
The data for XNLI can be downloaded with the following links and should be both saved (and un-zipped) in a $XNLI_DIR directory.

    XNLI 1.0
    XNLI-MT 1.0
```

Qbiwan commented:
> This is in good shape to be merged, thanks a lot for your work! I just have a few comments on how to simplify things here and there since there is only one task to deal with in the new script.
> 
> One question I have is, is the tokenizer the same for the training and evaluation datasets, even if the languages, can be different?

I'm puzzled. Is `Trainer()` class doing the magic under the hood when the languages are different? or is it `AutoTokenizer.from_pretrained`?

sgugger commented:
I missed this is a multinlingual checkpoint, so there is no need for different tokenizers.

Qbiwan commented:
@sgugger Yay :)
@patil-suraj let me know if there's anything you would like me to change further

patil-suraj reviewed (approved):
Looks good to me! Thank you for your work.

Qbiwan commented:
Thanks @sgugger @patil-suraj  for your helpful comments and guidance. I was jumping in at the deep end when I attempted this PR to be honest, but yay it's merged 😀

sgugger commented:
Great job adding this example and thanks a lot for your PR! Don't hesitate to brag a little bit on Twitter about your contribution ;-)

Khaliq88 commented on examples/text-classification/run_xnli.py:
```diff
@@ -16,611 +16,315 @@
+        return result
+
+    train_dataset = train_dataset.map(
+        preprocess_function, batched=True, load_from_cache_file=not data_args.overwrite_cache
```
#31971
